crates/randicheck/src/main.rs

```diff
@@ -0,0 +1,327 @@
+// A prototype haskell parser that can parse data types and functions with case statements
+use std::collections::HashMap;
+use streaming_iterator::StreamingIterator;
+use tree_sitter::{Parser, Query, QueryCursor, Tree};
+
+// a haskell data type has a name and a list of constructors
+#[derive(Debug)]
+struct DataType {
+    Name: String,
+    Constructors: Box<[Constructor]>,
+}
+
+//a constructor for an adt is just a list of types
+#[derive(Clone,Debug)]
+struct Constructor {
+    Types: Box<[Type]>,
+}
+// we only support bools and custom types for now (may have to add boolean with a T/F or placeholder later)
+#[derive(Clone,Debug)]
+enum Type {
+    Bool,
+    Custom(String),
+}
+
+// a function has a name, it's input type(s), and it's output type
+#[derive(Debug)]
+struct Function {
+    Name: String,
+    Inputs: Box<[Type]>,
+    Output: Box<Type>,
+    Body: Box<[FunctionBody]>,
+}
+//the body of a function is either a case statement which has what it is matching on and a pattern, or a boolean
+#[derive(Clone,Debug)]
+enum FunctionBody {
+    Case(String,Pattern),
+    Bool,
+}
+// a pattern has a list of types that it matches on and gives a boolean
+#[derive(Clone,Debug)]
+struct Pattern {
+    True_On: Box<[Type]>,
+    gives : Bool}
+
+#[derive(Clone,Debug)]
+enum Bool {
+    True,
+    False,
+}
+
+// the full haskell file is a list of data types and functions, where we only support custom types, bools, and functions with case statements
+#[derive(Debug)]
+struct Haskell {
+    adt: Box<[DataType]>,
+    fun: Box<[Function]>,
+}
+
+fn main() {
+    // get the haskell file
+    let code = include_str!("tst.hs");
+
+    // initialise parser
+    let mut parser = Parser::new();
+    let language = tree_sitter_haskell::LANGUAGE;
+    parser
+        .set_language(&language.into())
+        .expect("Error loading Haskell language");
+    // query to get the body of a function, which is a case statement, note that there is a recursion issue here at the apply function component
+    // as a function may be curried and we can't represent that in the current query language
+    let query_function = r#"
```

Comment:
I am still slightly worried about these queries. how do you know you consumed everything in the input?

Reply:
In short the plan is basically to purposely not do that, as we're only querying for the parts of haskell we can deal with right now, In terms of a more production style deployment, we can add or's with wildcards in areas to throw warnings/errors when potentially valuable input is skipped. eg for the function body we can add

``` [ <cases query> 
        _  @giveWarning] 
```

This is generally how I expect to handle recursion using queries as well. My main concern would probably be performance as I'm unsure how efficient queries are compared to manual traversal, but once using multiple languages it should be possible to nicely create language agnostic functions to pull data out of a rust enum or haskell adt just by loading a query cursor with the relevant query.